AppCenterAuth/AppCenterAuth/MSAuth.m

```diff
@@ -434,18 +462,62 @@ - (MSALAccount *)retrieveAccountWithAccountId:(NSString *)homeAccountId {
   return account;
 }
 
+- (void)refreshTokenForAccountId:(NSString *)accountId withNetworkConnected:(BOOL)networkConnected {
+  @synchronized(self) {
+    if (self.signInCompletionHandler) {
+      MSLogDebug([MSAuth logTag], @"Failed to refresh token: sign-in already in progress.");
+      return;
+    }
+    if (self.refreshCompletionHandler) {
+      MSLogDebug([MSAuth logTag], @"Token refresh already in progress. Skip this refresh request.");
+      return;
+    }
+    if (!networkConnected) {
+      MSLogDebug([MSAuth logTag], @"Network not connected. The token will be refreshed after coming back online.");
+      self.homeAccountIdToRefresh = accountId;
+      return;
+    }
+    MSALAccount *account = [self retrieveAccountWithAccountId:accountId];
+    if (account) {
+      __weak typeof(self) weakSelf = self;
+      self.refreshCompletionHandler = ^(MSUserInformation *_Nullable __unused userInformation, NSError *_Nullable __unused error) {
+        typeof(self) strongSelf = weakSelf;
+        @synchronized(strongSelf) {
+          strongSelf.refreshCompletionHandler = nil;
+        }
+      };
+      [self acquireTokenSilentlyWithMSALAccount:account
+                                     uiFallback:NO
+                    keyPathForCompletionHandler:NSStringFromSelector(@selector(refreshCompletionHandler))];
+    } else {
+
+      // If account not found, start an anonymous session to avoid deadlock.
+      MSLogWarning([MSAuth logTag],
+                   @"Could not get account for the accountId of the token that needs to be refreshed. Starting anonymous session.");
+      [[MSAuthTokenContext sharedInstance] setAuthToken:nil withAccountId:nil expiresOn:nil];
+    }
+  }
+}
+
 #pragma mark - MSAuthTokenContextDelegate
 
-- (void)authTokenContext:(MSAuthTokenContext *)authTokenContext refreshAuthTokenForAccountId:(nullable NSString *)accountId {
-  MSALAccount *account = [self retrieveAccountWithAccountId:accountId];
-  if (account) {
-    [self acquireTokenSilentlyWithMSALAccount:account];
-  } else {
+- (void)authTokenContext:(MSAuthTokenContext *)__unused authTokenContext refreshAuthTokenForAccountId:(nullable NSString *)accountId {
+  if (!accountId) {
+    return;
```

Comment:
We have this warning message here.
```
MSLogWarning([MSAuth logTag], @"Could not get account for the accountId of the token that needs to be refreshed. Starting anonymous session.");
```
If `accountId` is `nil`, it won't call `refreshTokenForAccountId:withNetworkConnected:` so the message won't be logged. You can just pass `accountId` and let the method handles it.